Add admin-only user-local setting that controls perms tutorial

escherize opened the pull request:
Fixes #45073 

Adds a modal and header to the data permissions pages with links to external documentation.

[Loom Demo](https://www.loom.com/share/0183b16e2331425d88b7801293e20f82)

[x] Tests have been added / updated

escherize committed:
add admin-only user-local setting that controls perms tutorial

npfitz committed:
Adding modal and banner, user setting for dismissable banner

npfitz committed:
copy changes

sloansparger commented on frontend/src/metabase/admin/permissions/pages/GroupDataPermissionsPage/GroupsPermissionsPage.jsx:
```diff
@@ -177,6 +182,11 @@ function GroupsPermissionsPage({
+              {showSplitPermsMessage && <PermissionsEditorSplitPermsMessage />}
+            </>
+          )}
           warnings={() => (
```
!nit messages and warnings being in different places feels weird. i'm wondering if we should rename the `messages` and `warnings` props to `preHeaderContent` and `postHeaderContent`.

sloansparger replied:
i still think this comment is relevant

npfitz replied:
Fair enough. I'll get it adjusted lol

sloansparger commented on frontend/src/metabase/admin/permissions/components/LegacyPermissionsModal/LegacyPermissionsModal.tsx:
```diff
@@ -0,0 +1,66 @@
+        </Modal.Header>
+        <Modal.Body p="2.5rem">
+          <Text mb="1rem">
+            {t`In Metabase 50, we split our data permissions into two new settings:`}{" "}
+            <Text span color="brand" fw="bold">{t`View data`}</Text> {t`and`}{" "}
+            <Text span color="brand" fw="bold">{t`Create queries`}</Text>
+            {`. Separate settings for what people can view and what they can query makes data permissions more expressive and easier to reason about. `}
```
I think you want to use the ttag method that allows you to use jsx. so that the first part of this string (line 40) and the last part (line 43) aren't separate for the translators.

(Also complete side note, but do we ever update the translations for a release? Won't this always be in english for v50?)

sloansparger reviewed (approved):
approved with comments

npfitz committed:
adjust so modal only appears on data permission page

npfitz committed:
e2e adjustments, adding banner test

escherize committed:
never show the modal or banner on new instances

escherize committed:
add test

noahmoss commented on src/metabase/public_settings.clj:
```diff
@@ -816,6 +816,48 @@ See [fonts](../configuring-metabase/fonts.md).")
                     ;; frontend should set this value to `true` after the modal has been shown once
                     v))))
 
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; v50 Permissions Tutorial settings
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
```
I think these should actually go into a different namespace. (Maybe `metabase.api.permissions`?) They're not public settings since they're admin-only, and I think they'd make more sense near the other perms-related code.

Sorry if I mislead with an earlier suggestion — I was just referring to moving the `instance-creation` setting which _is_ public.

escherize replied:
Makes sense, I'll move it there

noahmoss commented on src/metabase/public_settings.clj:
```diff
@@ -816,6 +816,48 @@ See [fonts](../configuring-metabase/fonts.md).")
+;; Bryan (7/8/24): These 2 settings are meant to mitigate any confusion over the new and improved permissions format
+;; We switched to in v50 for users who were around proir to v50. We never show the banner or modal on instances
+;; created after v50.
+;; We should come back around and delete them in a few months.
```
Should we actually _just_ merge this to the `release-x.50.x` branch but leave it out of master so that it's not in 51? Might be a product question for how long we want to leave this in. I guess some folks might skip 50 and upgrade straight to 51 from an earlier version so we could keep it in a few releases.

escherize replied:
I think I agree with what you mean, which is we should merge it to master, backport it, and revisit its existance in master in a couple months. I'll update the comment to make that clear

noahmoss commented on test/metabase/public_settings_test.clj:
```diff
@@ -370,3 +371,29 @@
+                                      [false false] false})
+
+(deftest show-updated-permission-banner-modal-test
+  (doseq [instance-creation-time [(t/local-date-time 2020) (t/local-date-time 2022)]
+          fifty-migration-time [(t/local-date-time 2021) (t/local-date-time 2023)]
+          banner-setting-value [true false]
+          modal-setting-value [true false]]
```
I'm personally in favor of `clojure.test/are` for this kind of enumeration. You can get something like
```
(are [instance-creation 50-migration banner-setting modal-setting] (test-fn ...)
  2020 2021 true true false
  2020 2021 true false false
  ...)
```
(or whatever the real values should be)
That way you can easily see the expected value for each set of inputs without tracing through the test logic 🤷 won't block on this but just an idea

escherize replied:
I actually thought that the map would be have less moving parts, since there are only 4 situations that make a difference, rather than writing out all 16 and making devs do the math to figure out what actually determines the output. The expected value calculations are listed above, in https://github.com/metabase/metabase/pull/45076/files#diff-d7af27f3905be1f9d80534697e89be5de2a5635586e3099ced9afb7def4da18eR375-R378.

escherize replied:
Maybe it would be easier to read if I make it into a map. I'll do that

npfitz committed:
e2e adjustments, copy update

sloansparger commented on e2e/test/scenarios/permissions/admin-permissions.cy.spec.js:
```diff
@@ -581,6 +581,51 @@ describe("scenarios > admin > permissions", () => {
+      });
+    }).as("sessionProps");
+
+    cy.intercept("api/setting/show-updated-permission-modal", () => {
+      tempState["show-updated-permission-modal"] = false;
+    });
+
+    cy.intercept("api/setting/show-updated-permission-banner", () => {
+      tempState["show-updated-permission-banner"] = false;
+    });
```
@npfitz can you add a brief comment on what this is doing? it's maybe not obvious that there's a second get request happening and you're altering the state at the time of these requests.

sloansparger commented on e2e/test/scenarios/permissions/download-permissions.cy.spec.js:
```diff
@@ -95,7 +95,6 @@ describeEE("scenarios > admin > permissions > data > downloads", () => {
 
   it("respects 'no download' permissions when 'All users' group data permissions are set to `Blocked` (metabase#22408)", () => {
     cy.visit(`/admin/permissions/data/database/${SAMPLE_DB_ID}`);
-
```
@npfitz great way to beef your stats

npfitz replied:
Don't tell @iethree I'm coming for his crown

sloansparger commented on frontend/src/metabase/admin/permissions/components/LegacyPermissionsModal/LegacyPermissionsModal.tsx:
```diff
@@ -0,0 +1,67 @@
+              <Text span color="brand" fw="bold">{t`View data`}</Text>
+            )} and ${(
+              <Text span color="brand" fw="bold">{t`Create queries`}</Text>
+            )} . Having separate settings for what people can view and what they can query makes data permissions more expressive and easier to reason about.`}
```
```suggestion
            )}. Having separate settings for what people can view and what they can query makes data permissions more expressive and easier to reason about.`}
```

sloansparger reviewed (approved):
LGTM, plz fix the small details i called out, but overall looks good

npfitz committed:
more PR feedback

escherize committed:
move modal/banner settings & tests + clarify tests

escherize committed:
Merge branch 'permissions-tutorial-modal' of github.com:metabase/metabase into permissions-tutorial-modal

escherize committed:
query for v50 migration execution works on mysql

- move changelog lookup into metabase.db.liquibase

escherize committed:
no race conditions in changelog-table-name

The previous implementation of this function was hitting this
liquibase issue: https://github.com/liquibase/liquibase/issues/2966

escherize committed:
revert changelog-table-name

- simply lookup the changlog row
